text/3451-additional-float-types.md

```diff
@@ -0,0 +1,139 @@
+- Feature Name: `additional-float-types`
+- Start Date: 2023-6-28
+- RFC PR: [rust-lang/rfcs#3451](https://github.com/rust-lang/rfcs/pull/3451)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+This RFC proposes new floating point types `f16` and `f128` into core language and standard library. Also this RFC introduces `f80`, `doubledouble`, `bf16` into `core::arch` for inter-op with existing native code.
+
+# Motivation
+[motivation]: #motivation
+
+IEEE-754 standard defines binary floating point formats, including binary16, binary32, binary64 and binary128. The binary32 and binary64 correspond to `f32` and `f64` types in Rust, while binary16 and binary128 are used in multiple scenarios (machine learning, scientific computing, etc.) and accepted by some modern architectures (by software or hardware).
+
+In C/C++ world, there're already types representing these formats, along with more legacy non-standard types specific to some platform. Introduce them in a limited way would help improve FFI against such code.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+`f16` and `f128` are primitive floating types, they can be used just like `f32` or `f64`. They always conform to binary16 and binary128 format defined in IEEE-754, which means size of `f16` is always 16-bit, and size of `f128` is always 128-bit.
+
+```rust
+let val1 = 1.0; // Default type is still f64
+let val2: f128 = 1.0;
+let val3: f16 = 1.0;
+let val4 = 1.0f128; // Suffix of f128 literal
+let val5 = 1.0f16; // Suffix of f16 literal
+
+println!("Size of f128 in bytes: {}", std::mem::size_of_val(&val2)); // 16
+println!("Size of f16 in bytes: {}", std::mem::size_of_val(&val3)); // 2
+```
+
+Because not every target supports `f16` and `f128`, compiler provides conditional guards for them:
+
+```rust
+#[cfg(target_has_f128)]
+fn get_f128() -> f128 { 1.0f128 }
+
+#[cfg(target_has_f16)]
+fn get_f16() -> f16 { 1.0f16 }
+```
+
+All operators, constants and math functions defined for `f32` and `f64` in core, are also defined for `f16` and `f128`, and guarded by respective conditional guards.
+
+`f80` type is defined in `core::arch::{x86, x86_64}`. `doubledouble` type is defined in `core::arch::{powerpc, powerpc64}`. `bf16` type is defined in `core::arch::{arm, aarch64, x86, x86_64}`. They do not have literal representation.
```

Comment:
```suggestion
The `f80` type is defined in `core::arch::{x86, x86_64}` as 80-bit extended precision. The `doubledouble`
type is defined in `core::arch::{powerpc, powerpc64}` and represent's PowerPC's non-IEEE double-double
format (two `f64`s used to aproximate `f128`). `bf16` type is defined in `core::arch::{arm, aarch64, x86, x86_64}` and  represents the "brain" float, a truncated `f32` with SIMD support on some hardware. These
types do not have literal representation.

When working with FFI, the `core::ffi::c_longdouble` type can be used to match whatever type
`long double` represents in C.
```

Reply:
Thanks, I did not add the mention of `longdouble` yet. More things need to be clarified:

- Is there always only 1 long double for each `(arch, abi, os)` tuple? For example, `powerpc64le-unknown-linux-gnu` can use either double or doubledouble or IEEE binary128 as `long double` by `-mabi=(ieee|ibm)longdouble` and `-mlong-double-(64|128)`.
- Is mangling of `long double` the same regardless of its underlying semantics?
- Some targets (also `powerpc64le` for example) support `.gnu_attribute`, so that linker can differentiate objects compiled by different long double ABI. Should Rust programs using `c_longdouble` emit such attribute?